src/Microsoft.AspNet.TestHost/WebSocketClient.cs

```diff
@@ -0,0 +1,177 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Net.Http;
+using System.Security.Cryptography;
+using Microsoft.Framework.Internal;
+using Microsoft.AspNet.Http.Features;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Http.Internal;
+
+namespace Microsoft.AspNet.TestHost
+{
+    public class WebSocketClient
+    {
+        private readonly Func<IFeatureCollection, Task> _next;
+        private readonly PathString _pathBase;
+
+        internal WebSocketClient([NotNull] Func<IFeatureCollection, Task> next, PathString pathBase)
+        {
+            _next = next;
+            
+            // PathString.StartsWithSegments that we use below requires the base path to not end in a slash.
+            if (pathBase.HasValue && pathBase.Value.EndsWith("/"))
+            {
+                pathBase = new PathString(pathBase.Value.Substring(0, pathBase.Value.Length - 1));
+            }
+            _pathBase = pathBase;
+
+            SubProtocols = new List<string>();
+        }
+
+        public IList<string> SubProtocols
+        {
+            get;
+            private set;
+        }
+
+        public Action<HttpRequestMessage> ConfigureRequest
+        {
+            get;
+            set;
+        }
+
+        public async Task<System.Net.WebSockets.WebSocket> ConnectAsync(Uri uri, CancellationToken cancellationToken)
+        {
+            // clientRequest
+            HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Get, uri);
+            request.Headers.Add("Connection", "Upgrade");
+            request.Headers.Add("Upgrade", "websocket");
+            request.Headers.Add("Sec-WebSocket-Version", "13");
+            request.Headers.Add("Sec-WebSocket-Key", CreateRequestKey());
+            if (SubProtocols.Count > 0)
+            {
+                request.Headers.Add("SecWebSocketProtocol", string.Join(", ", SubProtocols));
+            }
+            if (ConfigureRequest != null)
+            {
+                ConfigureRequest(request);
+            }
+
+            var state = new RequestState(request, _pathBase, cancellationToken);
+
+            // Async offload, don't let the test code block the caller.
+            var offload = Task.Factory.StartNew(async () =>
+            {
+                try
+                {
+                    await _next(state.FeatureCollection);
+                    state.PipelineComplete();
+                }
+                catch (Exception ex)
+                {
+                    state.PipelineFailed(ex);
+                }
+            });
+
+            return await state.WebSocketTask;
+        }
+
+        private class RequestState : IHttpWebSocketFeature
+        {
+            public IFeatureCollection FeatureCollection { get; private set; }
+            public Task<System.Net.WebSockets.WebSocket> WebSocketTask { get { return _websocketTcs.Task; } }
```

Comment:
Because of the name clash with WebSocket class in the Hosting namespace, will rename WebSocket to TestWebSocket as suggested by @Tratcher.